File: Backtracking/CombinationSumII.java
// LeetCode 40 - Combination Sum II
// Time Complexity: O(2^n) worst case | Space Complexity: O(target)
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSumII {
    private final List<List<Integer>> answer = new ArrayList<>();
    private final List<Integer> path = new ArrayList<>();
    private int[] candidates;

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        this.candidates = candidates;
        dfs(0, target);
        return answer;
    }

    private void dfs(int start, int remain) {
        if (remain == 0) {
            answer.add(new ArrayList<>(path));
            return;
        }
        if (start >= candidates.length || remain < candidates[start]) {
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            if (i > start && candidates[i] == candidates[i - 1]) {
                continue;
            }

            path.add(candidates[i]);
            dfs(i + 1, remain - candidates[i]);
            path.remove(path.size() - 1);
        }
    }
}

File: Backtracking/CombinationSumII.go
// LeetCode 40 - Combination Sum II
// Time Complexity: O(2^n) worst case | Space Complexity: O(target)
package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	answer := make([][]int, 0)
	path := make([]int, 0)

	var dfs func(start, remain int)
	dfs = func(start, remain int) {
		if remain == 0 {
			combo := make([]int, len(path))
			copy(combo, path)
			answer = append(answer, combo)
			return
		}
		if start >= len(candidates) || remain < candidates[start] {
			return
		}

		for i := start; i < len(candidates); i++ {
			if i > start && candidates[i] == candidates[i-1] {
				continue
			}
			path = append(path, candidates[i])
			dfs(i+1, remain-candidates[i])
			path = path[:len(path)-1]
		}
	}

	dfs(0, target)
	return answer
}
